### PR TITLE
CBMC: Add spec + proof for rej_uniform

### LINKED ISSUE #316: CBMC: State & prove specs for rej_uniform

### PR DESCRIPTION
**Pull request by hanno-becker**
Resolves #316

[//]: # (SPDX-License-Identifier: CC-BY-4.0)

### CODE REVIEW

File: mlkem/rej_uniform.c [hunk 1]
```diff
@@ -12,10 +12,11 @@
  *
  * Arguments:   - int16_t *r:          pointer to output buffer
  *              - unsigned int len:    requested number of 16-bit integers
- *                                     (uniform mod q)
+ *                                     (uniform mod q) Must be <= 1024.
  *              - const uint8_t *buf:  pointer to input buffer
  *                                     (assumed to be uniform random bytes)
  *              - unsigned int buflen: length of input buffer in bytes
+ *                                     Must be <= 4096.
```
Comment (mkannwischer) [inline mlkem/rej_uniform.c:19]: I don't understand where this limitation is coming from? Is this a CBMC limitation? In that case I would not make it part of the docstring here as this is not a real limitation of the C code. 
You can instead add a not to the `REQUIRES` stmt to clarify why that pre-condition exists.

File: mlkem/rej_uniform.h [hunk 2]
```diff
@@ -2,8 +2,35 @@
...
+ **************************************************/
 unsigned int rej_uniform(int16_t *r, unsigned int len, const uint8_t *buf,
-                         unsigned int buflen);
-
+                         unsigned int buflen)  // clang-format off
+REQUIRES(buflen <= 4096 && len <= 1024)
+REQUIRES(r != NULL && IS_FRESH(r, sizeof(int16_t) * len))
+REQUIRES(buf != NULL && IS_FRESH(buf, buflen))
+ASSIGNS(OBJECT_UPTO(r, len * sizeof(int16_t)))
+ENSURES(RETURN_VALUE <= len && \
```
Comment (mkannwischer) [inline mlkem/rej_uniform.h:33]:
```suggestion
ENSURES(RETURN_VALUE <= len && 
```

**Review by mkannwischer — CHANGES_REQUESTED** [review]
Thanks @hanno-becker. 
Bounds do look good.
I have one questions regarding the input size limitation that probably requires the addition of a comment.

File: mlkem/rej_uniform.c (hunk 1 above)
Comment (hanno-becker) [reply mlkem/rej_uniform.c:19]: This is not a CBMC limitation -- the `pos + 3` calculation can overflow if you have `len == UINT_MAX-2`. I mentioned that in the code.

I don't like getting the doc and spec get out of sync

Comment (hanno-becker) [reply mlkem/rej_uniform.c:19]: To be fair, the limitation is for `buflen` only, not `inlen`. 

If you prefer, I can state the spec as strong as possible and give the largest possible bound for `buflen`.

File: mlkem/rej_uniform.h (hunk 2 above)
Comment (hanno-becker) [reply mlkem/rej_uniform.h]: done

File: mlkem/rej_uniform.c (hunk 1 above)
Comment (hanno-becker) [reply mlkem/rej_uniform.c]: Ok, left a comment and removed the bound for `len`

Comment (mkannwischer) [reply mlkem/rej_uniform.c]: Thanks for adjusting this!

**Review by mkannwischer — APPROVED** [review]
Thanks for your edits! LGTM.

**Pull request merged**